### PR TITLE
PICSA Rainfall graphs dialog: Added facet options and bug fixes

### LINKED ISSUE #6106: New bug on PICSA rainfall graph
[four_stations_raindays.zip](https://github.com/africanmathsinitiative/R-Instat/files/5602352/four_stations_raindays.zip)

In the graph above I used multiple y of the 3 levels of raindays.  The Name (Station) was a facet, but I don't think that is relevant.

It works fine until I add a line for the mean.  Then it gives an error as follows:
image

### PR DESCRIPTION
**Pull request by N-thony**
Fixes #6106

### CODE REVIEW

**Comment by dannyparsons** [comment]
@N-thony I have edited the title of the PR so that's its more specific and descriptive. We try to use the PR titles in the release notes so please make them descriptive by mentioning the dialog name and the changes in a few words.

**Comment by N-thony** [comment]
> @N-thony I have edited the title of the PR so that's its more specific and descriptive. We try to use the PR titles in the release notes so please make them descriptive by mentioning the dialog name and the changes in a few words.

Thanks

**Comment by N-thony** [comment]
Added pull-down control for facet options and bug fixes. This solves issue #6106
@africanmathsinitiative/developers and @rdstern this is ready for review. Thanks.

**Comment by lloyddewit** [comment]
@Ivanluv please could you review? thanks

**Comment by Ivanluv** [comment]
reviewing this now

File: instat/dlgPICSARainfall.vb [hunk 1]
```diff
@@ -615,6 +658,129 @@ Public Class dlgPICSARainfall
...
+                        bRow = True
+                    End If
+                End If
+            Next
+            If bWrap OrElse bRow OrElse bCol Then
+                clsBaseOperator.AddParameter("facets", clsRFunctionParameter:=clsFacetFunction)
+            End If
+            If bWrap Then
+                clsFacetFunction.SetRCommand("facet_wrap")
+                clsFacetOperator.AddParameter("wrap", iPosition:=0)
```
Comment (Ivanluv) [inline instat/dlgPICSARainfall.vb:761]: This  parameter  does not to have a parameter value

**Comment by N-thony** [comment]
@Ivanluv I solved the issue you raised minutes ago. Could you confirm that the issue is resolved? Thanks.

File: instat/dlgPICSARainfall.vb [hunk 2]
```diff
@@ -678,7 +841,7 @@ Public Class dlgPICSARainfall
     End Sub
 
     Private Sub ucrReceiverX_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrReceiverX.ControlValueChanged
-        XAxisDataTypeCheck()
+        ' XAxisDataTypeCheck()
     End Sub
 
     Private Sub XAxisDataTypeCheck()
```
Comment (Ivanluv) [inline instat/dlgPICSARainfall.vb:847]: This function is also not being referenced anywhere in you code. Do you still need it?

File: instat/dlgPICSARainfall.vb [hunk 3]
```diff
@@ -742,8 +906,7 @@ Public Class dlgPICSARainfall
             End If
 
             If Not ucrVariablesAsFactorForPicsa.bSingleVariable Then
-                clsGroupByFunction.AddParameter(i, "variable", bIncludeArgumentName:=False, iPosition:=0)
-                i = i + 1
+                clsGroupByFunction.AddParameter("variable", bIncludeArgumentName:=False, iPosition:=0)
```
Comment (Ivanluv) [inline instat/dlgPICSARainfall.vb:909]: this parameter does not have a parameter values

File: instat/dlgPICSARainfall.vb [hunk 4]
```diff
@@ -678,7 +841,7 @@ Public Class dlgPICSARainfall
     End Sub
 
     Private Sub ucrReceiverX_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrReceiverX.ControlValueChanged
```
Comment (Ivanluv) [inline instat/dlgPICSARainfall.vb:843]: could you remove this sub since its no longer being used

**Comment by N-thony** [comment]
@Ivanluv I solved the issues you raised minutes ago. Could you confirm that the issues are resolved? Thanks.

**Comment by Ivanluv** [comment]
@lloyddewit the code looks okay from my end

**Comment by lloyddewit** [comment]
@rdstern please could you test? thanks

**Review by rdstern — APPROVED** [review]
Great.  Both the enhancement and the addition of mean lines seems to work fine!

File: instat/dlgPICSARainfall.vb (hunk 3 above)
Comment (maxwellfundi) [reply instat/dlgPICSARainfall.vb:907]: I am concerned about this parameter that has no value, if its not needed can it be removed? 
While @rdstern  has tested this and approved it, when should this parameter be added @N-thony ?

**Comment by rdstern** [comment]
@maxwellfundi I just comment on our scheme, because you are doing a perfect job here!  I only test the functionality and not the code.  From my point of view it seemed to work ok.

Then @lloyddewit or shadrack or you, or Danny  check the code.  That may be before me or after me.  If you find problems then the pull request changes and my approval becomes null.  I then look again after your suggestions are satisfied.

File: instat/dlgPICSARainfall.vb (hunk 3 above)
Comment (N-thony) [reply instat/dlgPICSARainfall.vb:907]: @maxwellfundi the bug was due the duplication of this parameter variable in the group_by function when you have more than 1 station

File: instat/dlgPICSARainfall.vb [hunk 5]
```diff
@@ -76,6 +80,17 @@ Public Class dlgPICSARainfall
     Private clsGeomHlineAesLowerTercile As New RFunction
     Private clsGeomHlineAesUpperTercile As New RFunction
 
+    Private strFacetWrap As String = "Facet Wrap"
+    Private strFacetRow As String = "Facet Row"
+    Private strFacetCol As String = "Facet Column"
+    Private strXAxis As String = "X Axis"
+    Private strColour As String = "Colour Axis"
+    Private strNone As String = "None"
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:88]: Make all these strings `ReadOnly`?

File: instat/dlgPICSARainfall.vb [hunk 6]
```diff
@@ -615,6 +658,126 @@ Public Class dlgPICSARainfall
         End If
     End Sub
 
+    Private Sub ucrInput_ControlValueChanged(ucrChangedControl As ucrInputComboBox) Handles ucrInputStation.ControlValueChanged
+        Dim ucrTempReciever As ucrReceiver = Nothing
+        Dim strChangedText As String = ""
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:663]:
```suggestion
```

File: instat/dlgPICSARainfall.vb [hunk 7]
```diff
@@ -615,6 +658,126 @@ Public Class dlgPICSARainfall
         End If
     End Sub
 
+    Private Sub ucrInput_ControlValueChanged(ucrChangedControl As ucrInputComboBox) Handles ucrInputStation.ControlValueChanged
+        Dim ucrTempReciever As ucrReceiver = Nothing
+        Dim strChangedText As String = ""
+
+        If bUpdateComboOptions Then
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:665]: Follow the 'happy path stays left` principle (https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88). So more readable if we write:
```
        If Not bUpdateComboOptions Then
            Exit Sub
        End If
```

File: instat/dlgPICSARainfall.vb [hunk 8]
```diff
@@ -615,6 +658,126 @@ Public Class dlgPICSARainfall
...
+                    End If
+                Next
+            End If
+            UpdateParameters()
+            AddRemoveFacets()
+        End If
+    End Sub
+
+    Private Sub UpdateParameters()
+        Dim strTemp As String = ""
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:685]:
```suggestion
```

File: instat/dlgPICSARainfall.vb [hunk 9]
```diff
@@ -615,6 +658,126 @@ Public Class dlgPICSARainfall
...
+        bUpdatingParameters = False
+    End Sub
+
+    Private Sub AddRemoveFacets()
+        Dim strText As String
+        Dim bWrap As Boolean = False
+        Dim bCol As Boolean = False
+        Dim bRow As Boolean = False
+
+        If Not bUpdatingParameters Then
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:740]: Invert condition so that 'happy path stays left'

File: instat/dlgPICSARainfall.vb [hunk 10]
```diff
@@ -76,6 +80,17 @@ Public Class dlgPICSARainfall
...
+    Private strFacetWrap As String = "Facet Wrap"
+    Private strFacetRow As String = "Facet Row"
+    Private strFacetCol As String = "Facet Column"
+    Private strXAxis As String = "X Axis"
+    Private strColour As String = "Colour Axis"
+    Private strNone As String = "None"
+
+    Private bUpdateComboOptions As Boolean = True
+    Private bUpdatingParameters As Boolean = False
+    Private dctComboReceiver As New Dictionary(Of ucrInputComboBox, ucrReceiverSingle)
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:92]: This `Dictionary` only contains a single value. Could we remove this data member and just reference to `ucrReceiverFacetBy` instead?

File: instat/dlgPICSARainfall.vb [hunk 11]
```diff
@@ -144,14 +160,16 @@ Public Class dlgPICSARainfall
...
+        ucrReceiverFacetBy.SetParameter(New RParameter(""))
         ucrReceiverFacetBy.Selector = ucrSelectorPICSARainfall
         ucrReceiverFacetBy.SetIncludedDataTypes({"factor"})
         ucrReceiverFacetBy.strSelectorHeading = "Factors"
         ucrReceiverFacetBy.bWithQuotes = False
         ucrReceiverFacetBy.SetParameterIsString()
         ucrReceiverFacetBy.SetValuesToIgnore({"."})
 
+        dctComboReceiver.Add(ucrInputStation, ucrReceiverFacetBy)
+
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:172]: Remove this line?

File: instat/dlgPICSARainfall.vb [hunk 12]
```diff
@@ -615,6 +658,126 @@ Public Class dlgPICSARainfall
...
     End Sub
 
+    Private Sub ucrInput_ControlValueChanged(ucrChangedControl As ucrInputComboBox) Handles ucrInputStation.ControlValueChanged
+        Dim ucrTempReciever As ucrReceiver = Nothing
+        Dim strChangedText As String = ""
+
+        If bUpdateComboOptions Then
+            strChangedText = ucrChangedControl.GetText()
+            If strChangedText <> strNone Then
+                For Each ucrInputTemp As ucrInputComboBox In dctComboReceiver.Keys
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:668]: The dictionary only has one entry. Remove this loop?

File: instat/dlgPICSARainfall.vb [hunk 13]
```diff
@@ -615,6 +658,126 @@ Public Class dlgPICSARainfall
...
+
+        For Each ucrInputTemp As ucrInputComboBox In dctComboReceiver.Keys
+            clsFacetOperator.RemoveParameterByName("wrap" & ucrInputTemp.Name)
+            clsFacetColOp.RemoveParameterByName("col" & ucrInputTemp.Name)
+            clsFacetRowOp.RemoveParameterByName("row" & ucrInputTemp.Name)
+        Next
+        clsBaseOperator.RemoveParameterByName("facets")
+
+        bUpdatingParameters = True
+        For Each ucrInputTemp As ucrInputComboBox In dctComboReceiver.Keys
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:700]: The dictionary only has one entry. Remove this loop?

File: instat/dlgPICSARainfall.vb [hunk 14]
```diff
@@ -615,6 +658,126 @@ Public Class dlgPICSARainfall
...
+
+    Private Sub AddRemoveFacets()
+        Dim strText As String
+        Dim bWrap As Boolean = False
+        Dim bCol As Boolean = False
+        Dim bRow As Boolean = False
+
+        If Not bUpdatingParameters Then
+            clsBaseOperator.RemoveParameterByName("facets")
+            For Each kvpTemp As KeyValuePair(Of ucrInputComboBox, ucrReceiverSingle) In dctComboReceiver
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:742]: The dictionary only has one entry. Remove this loop?

File: instat/dlgPICSARainfall.vb [hunk 15]
```diff
@@ -615,6 +658,126 @@ Public Class dlgPICSARainfall
...
 
+    Private Sub ucrInput_ControlValueChanged(ucrChangedControl As ucrInputComboBox) Handles ucrInputStation.ControlValueChanged
+        Dim ucrTempReciever As ucrReceiver = Nothing
+        Dim strChangedText As String = ""
+
+        If bUpdateComboOptions Then
+            strChangedText = ucrChangedControl.GetText()
+            If strChangedText <> strNone Then
+                For Each ucrInputTemp As ucrInputComboBox In dctComboReceiver.Keys
+                    If Not strChangedText = strFacetCol AndAlso Not strChangedText = strFacetRow AndAlso Not ucrInputTemp.Equals(ucrChangedControl) AndAlso ucrInputTemp.GetText() = strChangedText Then
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:669]: make more readable by adding new lines between conditions

File: instat/dlgPICSARainfall.vb [hunk 16]
```diff
@@ -615,6 +658,126 @@ Public Class dlgPICSARainfall
...
+                        bUpdateComboOptions = False
+                        ucrInputTemp.SetName(strNone)
+                        bUpdateComboOptions = True
+                    End If
+                    If strChangedText = strFacetWrap AndAlso ucrInputTemp.GetText = strFacetRow OrElse strChangedText = strFacetRow AndAlso ucrInputTemp.GetText = strFacetWrap OrElse strChangedText = strFacetWrap AndAlso ucrInputTemp.GetText = strFacetCol OrElse strChangedText = strFacetCol AndAlso ucrInputTemp.GetText = strFacetWrap Then
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:674]: make more readable by adding new lines between conditions.
Also, the mix between `AndAlso` and `OrElse` is confusing. Add brackets to clarify the precedence?

File: instat/dlgPICSARainfall.vb [hunk 17]
```diff
@@ -615,6 +658,126 @@ Public Class dlgPICSARainfall
...
+
+    Private Sub UpdateParameters()
+        Dim strTemp As String = ""
+
+        'clsRaesFunction.RemoveParameterByName("x")
+        clsRaesFunction.RemoveParameterByName("color")
+        clsRaesFunction.RemoveParameterByName("fill")
+        clsAsFactorFunction.RemoveParameterByName("x")
+
+        For Each ucrInputTemp As ucrInputComboBox In dctComboReceiver.Keys
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:692]: The dictionary only has one entry. Remove this loop?

File: instat/dlgPICSARainfall.vb [hunk 18]
```diff
@@ -615,6 +658,126 @@ Public Class dlgPICSARainfall
...
+            clsFacetColOp.RemoveParameterByName("col" & ucrInputTemp.Name)
+            clsFacetRowOp.RemoveParameterByName("row" & ucrInputTemp.Name)
+        Next
+        clsBaseOperator.RemoveParameterByName("facets")
+
+        bUpdatingParameters = True
+        For Each ucrInputTemp As ucrInputComboBox In dctComboReceiver.Keys
+            strTemp = ucrInputTemp.GetText()
+            dctComboReceiver(ucrInputTemp).SetRCode(Nothing)
+            If strTemp = strXAxis Then
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:703]: Replace this 'ElseIf` structure with `Select`

File: instat/dlgPICSARainfall.vb [hunk 19]
```diff
@@ -615,6 +658,126 @@ Public Class dlgPICSARainfall
...
+            End If
+        Next
+        'If Not clsRaesFunction.ContainsParameter("x") Then
+        '    clsRaesFunction.AddParameter("x", Chr(34) & Chr(34))
+        'End If
+        bUpdatingParameters = False
+    End Sub
+
+    Private Sub AddRemoveFacets()
+        Dim strText As String
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:735]: delete?

File: instat/dlgPICSARainfall.vb [hunk 20]
```diff
@@ -615,6 +658,126 @@ Public Class dlgPICSARainfall
...
+                If strText = strFacetWrap OrElse strText = strFacetCol OrElse strText = strFacetRow AndAlso Not kvpTemp.Value.IsEmpty Then
+                    If strText = strFacetWrap Then
+                        bWrap = True
+                    End If
+                    If strText = strFacetCol Then
+                        bCol = True
+                    End If
+                    If strText = strFacetRow Then
+                        bRow = True
+                    End If
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:753]: replace with select

File: instat/dlgPICSARainfall.vb [hunk 21]
```diff
@@ -615,6 +658,126 @@ Public Class dlgPICSARainfall
...
+        Dim strText As String
+        Dim bWrap As Boolean = False
+        Dim bCol As Boolean = False
+        Dim bRow As Boolean = False
+
+        If Not bUpdatingParameters Then
+            clsBaseOperator.RemoveParameterByName("facets")
+            For Each kvpTemp As KeyValuePair(Of ucrInputComboBox, ucrReceiverSingle) In dctComboReceiver
+                strText = kvpTemp.Key.GetText()
+                If strText = strFacetWrap OrElse strText = strFacetCol OrElse strText = strFacetRow AndAlso Not kvpTemp.Value.IsEmpty Then
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:744]: remove first 3 conditions

File: instat/dlgPICSARainfall.vb (hunk 3 above)
Comment (lloyddewit) [reply instat/dlgPICSARainfall.vb:907]: @N-thony Sorry, I still don't understand why this parameter does not have a value, please could you explain? or we can quickly discuss verbally) thanks

**Review by lloyddewit — CHANGES_REQUESTED** [review]

**Comment by N-thony** [comment]
@maxwellfundi @lloyddewit thanks for your comments above. I think to have resolved them.

**Comment by lloyddewit** [comment]
@N-thony thank you for the changes. However, I still see a number of unresolved comments. Please resolve and let me know when you are done.
If my comments are not clear then let me know and we can discuss.
Thanks

File: instat/dlgPICSARainfall.vb (hunk 19 above)
Comment (N-thony) [reply instat/dlgPICSARainfall.vb:750]: @lloyddewit we can discuss.

**Comment by N-thony** [comment]
@lloyddewit thanks for your comments. I think have resolved them.

**Comment by N-thony** [comment]
@lloyddewit thanks for your comments. I think have resolved them.

File: instat/dlgPICSARainfall.vb [hunk 22]
```diff
@@ -615,6 +651,119 @@ Public Class dlgPICSARainfall
...
+                    dctComboReceiver(ucrInputTemp).ChangeParameterName("wrap" & ucrInputTemp.Name)
+                    dctComboReceiver(ucrInputTemp).SetRCode(clsFacetOperator)
+                Case strFacetCol
+                    dctComboReceiver(ucrInputTemp).ChangeParameterName("col" & ucrInputTemp.Name)
+                    dctComboReceiver(ucrInputTemp).SetRCode(clsFacetColOp)
+                Case strFacetRow
+                    dctComboReceiver(ucrInputTemp).ChangeParameterName("row" & ucrInputTemp.Name)
+                    dctComboReceiver(ucrInputTemp).SetRCode(clsFacetRowOp)
+            End Select
+        Next
```
Comment (lloyddewit) [inline instat/dlgPICSARainfall.vb:715]: Does this replacement code work and allow you to remove `dctComboReceiver`?
```suggestion
        ucrReceiverFacetBy.SetRCode(Nothing)
        Select Case ucrInputStation.GetText()
            Case strXAxis
                ucrReceiverFacetBy.ChangeParameterName("x")
                ucrReceiverFacetBy.SetParameterIncludeArgumentName(False)
                ucrReceiverFacetBy.SetRCode(clsAsFactorFunction)
                clsRaesFunction.AddParameter("x", clsRFunctionParameter:=clsAsFactorFunction)
            Case strColour
                ucrReceiverFacetBy.ChangeParameterName("color")
                ucrReceiverFacetBy.SetParameterIncludeArgumentName(True)
                ucrReceiverFacetBy.SetRCode(clsRaesFunction)
            Case strFacetWrap
                ucrReceiverFacetBy.ChangeParameterName("wrap" & ucrInputStation.Name)
                ucrReceiverFacetBy.SetRCode(clsFacetOperator)
            Case strFacetCol
                ucrReceiverFacetBy.ChangeParameterName("col" & ucrInputStation.Name)
                ucrReceiverFacetBy.SetRCode(clsFacetColOp)
            Case strFacetRow
                ucrReceiverFacetBy.ChangeParameterName("row" & ucrInputStation.Name)
                ucrReceiverFacetBy.SetRCode(clsFacetRowOp)
        End Select
```

File: instat/dlgPICSARainfall.vb (hunk 10 above)
Comment (lloyddewit) [reply instat/dlgPICSARainfall.vb:92]: Please see comment on line 688

File: instat/dlgPICSARainfall.vb (hunk 11 above)
Comment (lloyddewit) [reply instat/dlgPICSARainfall.vb:172]: Please see comment on line 688

**Review by lloyddewit — CHANGES_REQUESTED** [review]

**Comment by N-thony** [comment]
@lloyddewit thanks for your comment. I removed the dictionary and it works.

**Review by lloyddewit — DISMISSED** [review]

**Comment by lloyddewit** [comment]
@rdstern There have been quite a few code changes since your last test, please could you test again? thanks

**Review by rdstern — CHANGES_REQUESTED** [review]
@N-thony the main issue of the bug seems to have gone.  Great.
1. In the options for the station to be a facet I think you can delete x-axis and probably also colour axis.
2. Try - with the sample data set mentioned - to change the option for the trpe of facet.  I made it Wrap.  Then it also deleted the year from the x-axis.  They should not be connected in this way.

**Comment by maxwellfundi** [comment]
@N-thony there requested changes here, could you make these a priority?

**Comment by N-thony** [comment]
> @N-thony there requested changes here, could you make these a priority?

Okay @maxwellfundi. I need to check this first with @rdstern.

**Comment by N-thony** [comment]
@rdstern thanks for your comment above. I think have resolved it. Could you test the change?

**Review by rdstern — APPROVED** [review]
@lloyddewit the problems I had are now fixed.

**Review by lloyddewit — APPROVED** [review]

**Pull request merged**